File: leetcode/2. problem/70.爬楼梯.java
/*
 * @lc app=leetcode.cn id=70 lang=java
 *
 * [70] 爬楼梯
 */

// @lc code=start
class Solution {

    public int climbStairs(int n) {

        // 方法一 动态规划1 最优解->去掉了不必要的数据
        // return extracted(n);

        // 方法二 动态规划2 : 保存了所有的计算值，空间上会有一些浪费
        // return extracted2(n);

        // test
        return test(n);

    }

    private int test(int n) {
        int[] dp = new int[n + 1];
        // 第一级台阶有一种方法
        dp[1] = 1;
        if (n < 2) {
            return 1;
        }
        // 第二级台阶有两种方法
        dp[2] = 2;
        for (int i = 3; i <= n; i++) {
            dp[i] = dp[i - 1] + dp[i - 2];
        }
        return dp[n];
    }

    /**
     * 动态规划2
     * 
     * Your runtime beats 100 % of java submissions
     * Your memory usage beats 23.91 % of java submissions (38.2 MB)
     * 
     * @param n
     * @return
     */
    private int extracted2(int n) {

        int[] dp = new int[n + 1];
        dp[0] = 1;
        dp[1] = 1;
        for (int i = 2; i <= n; i++) {
            dp[i] = dp[i - 1] + dp[i - 2];
        }
        return dp[n];
    }

    /**
     * 动态规划1
     * 
     * Your runtime beats 100 % of java submissions
     * Your memory usage beats 32.87 % of java submissions (38.1 MB)
     * 
     * @param n
     * @return
     */
    private int extracted(int n) {
        // f(x) = f(x-1)+f(x-2)
        // 动态规划
        int p = 0, q = 0, r = 1;
        for (int i = 0; i < n; i++) {
            p = q;
            q = r;
            r = p + q;
        }
        return r;
    }
}
// @lc code=end

File: leetcode/2. problem/70.爬楼梯.go
/*
 * @lc app=leetcode.cn id=70 lang=golang
 *
 * [70] 爬楼梯
 */

package main

// @lc code=start
func climbStairs(n int) int {
	return newFunction3(n)
}

/*
45/45 cases passed (0 ms)
Your runtime beats 100 % of golang submissions
Your memory usage beats 41.27 % of golang submissions (1.8 MB)
*/
func newFunction3(n int) int {
	dp := make([]int, n+1)
	dp[1] = 1
	if n < 2 {
		return 1
	}
	dp[2] = 2
	for i := 3; i <= n; i++ {
		dp[i] = dp[i-1] + dp[i-2]
	}
	return dp[n]
}

// @lc code=end
